,Update turkish tables (tables/tr.ctb, tables/tr-g2.ctb)

hammera opened the pull request:
Hi Boys,

@oguzugur28 and he's author friend doed some larger modifications with tables/tr.ctb, tables/tr-g2.ctb files, and sent me and oldest time the Liblouis list the table files modifications.
I promised Oguz with I opening a pull request with this changes related.

tables/tr.ctb file happened following changes:
Oğuz UĞUR and Ozan UĞUR added with various updates (e-mail adresses awailable the tables).
Following informations are added with Oguz, i quote the commit message with history purpose:
Punctuation symbols is obtained from Turkish Braille Writing Guideline released by Turkish Education Ministery in 1991. For unaccessible PDF version see:
https://orgm.meb.gov.tr/alt_sayfalar/yayimlar/ozelegitim/braille/braille.pdf
Contrary to Official Guide, In the Grade 2 Table, for providing readibilty, some accented letters' dot combination entries were deleted and these letters' entries are presented similar
other accented letters. For example, some accented letters presented with dot 26 or 1246 but, also these  dot combinations are used to present some abbreviations, therefore, to prevent
mixing accented letters with abbreviations, these letters' dot combinations in Turkish Grade 2 table were deleted.

With grade2 table I think happened following changes, quote from Oguz e-mail:
First, I have corrected abriviation mistakes. Some abriviation patterns were added wrong, I have deleted it and added correct ones.

Also I have changed some abriviations’ rules. Some abriviations which are called "One Letter Abriviations" only must located begining of words but current official Liblouis releases, this rule isn’t applied on Turkish tables. I corrected it. For example, dots 1234, which is P, mean -para- in Turkish and para mean money in English. -İmparator- mean emperor in English, and as well you see, İmparator word includes -para-. According to Turkish braille rules, -imparator- word musn't include -para- abriviation. Because -para- is located middle of word. To correct presentation, I have added begmidword code before the -para- abriviation and other all One Letter Abriviations.

 

Also, for released Liblouis Turkish tables, punctuation list for Turkish is incompleted. I have added many of them.

 

Finally, Turkish includes accented letters. I have followed different way to solve this problem. I have derived accented letters from German Accented Letters file. Thus, at least with my work, many foreign origined accented letters are presented correctly now.

Oguz have Windows 11 operating system, JAWS' most current version and Orbit Reader 40 Braille display (this is the testing possibilities he's machine).

Bert, Chris, don't surprise, I not yet doed the testcase files from Oguz doed changes (me need analyze the diff outputs to generate only the changed rules a first turkish testcase file), but make check commands rans succesfull with the tests directory.

Because now Oguz not use Git yet, I am have possibilities to correct problems if you founding any change during review, when Oguz sending me e-mail for example the proper changed versions if need changing anything, I simple committing my forked repository the Oguz changed tables.
So, I am only the technical contributor (committer person), real author the new changes with Oguz and he's friend with are defined the table files:
* Oğuz UĞUR <oguzugur2834@gmail.com>,
* Ozan UĞUR <ozanugur28@icloud.com>

Attila

oguzugur28 commented:
Hi All, Best wishes from Turkey. I have explained Attila and Bert, I'm not a software developer or code contributor but I'm using braille in my daily life. Therefore, I have started to update Turkish tables. Thanks to Attila, Attila has accepted my support request.

hammera commented:
Ok boys, only remaining the tr-g2.ctb table and testcase related work.
With @oguzugur28 doed fantastic proofread work and sent me corrections with tables/tr-g1.ctb and tests/braille-specs/tr-g1_harness.yaml file related.
I integrated the final tr-g1_harness.yaml testcase file the build system.

Before when we merging the final reviewed work, we need only hide the „fixup” type commits with a simple git -i rebase and a git force-push command.
Perhaps I will hide this commits when the main branch are changed an another PR related. :-):-)

Oguz, You and your friends are doed fantastic work, I congratulation for you and your friends!

Attila

hammera commented:
Ok, main tests/braille-specs/tr-g2_harness.yaml file is ready and integrated the build system.
I waiting from Oguz only with 34 begword, begmidword or sufword related real examples to test new abbreviations the yaml file, other testcases are Oguz fantastic proofreaded and verifyed yesterday evening.
I looked only afternoon the tr-g2.ctb table related commit to I not forgot any new element from the testcase file.

So, when I get the 35 example words, only completing the abbreviations examples the tests/braille-specs/tr-g2_harness.yaml file a new commit.
I hided with fixup commits with a git rebase and a force-push operation.
After I pasted Oguz real word examples the tr-g2_harness.yaml file, the pull request are review ready my openion, I done the development and pre-review related task. :-):-)

Attila

egli commented on tables/tr-g1.ctb:
```diff
@@ -30,19 +30,179 @@
+lowercase x 4-1346
+# Turkish characters (added by Arend Arends) and Oguz Ogur
 
-include en-ueb-g1.ctb
```
Why did you remove this include? Was it so that you can overwrite the basic latin capital i? Could you not keep the include and overwrite the definition of the i in some other way?

hammera replied:
Hi Chris,

Yes, exactly, the different i letters related rule needs I need moved the include line from top of tr-g1.ctb table to end of tr-g1.ctb table.
When I experimented with entering different variations of the letter i so that the include line was at the beginning of the table in the original place, the correct order of precedence simply did not take effect, the point changes requested by Oguz did not take effect. However, according to the read grade1 test result, all characters are good in the case of the tr-g1.ctb table, this was also confirmed by Oguz.
Of course this is true only when the include line is have the end of the tr-g1.ctb table.
If I put a hassmark the include en-ueb-g1.ctb line, the tr-g1.yaml produce 155 failure results.
If I putting back the 38TH line the include en-ueb-g1.ctb line, the grade1 turkish test file produce I think 168 failures.
So, I am fraid this is the best combination (keeping the end of table the include en-ueb-g1.ctb include line).
This way when I requested Oguz the grade 1 test entire proofread, he says all dot combination is right if we use this variation.

egli replied:
OK, sounds good

egli commented on tests/braille-specs/Makefile.am:
```diff
@@ -123,6 +123,9 @@ EXTRA_DIST =					\
 	sw-ke.yaml				\
 	ta-ta-g1_harness.yaml			\
 	th.yaml					\
+	tr-8dots_harness.yaml					\
+		tr-g1_harness.yaml					\
+		tr-g2_harness.yaml					\
```
I prefer that the names of the yaml files to be just `tr-8dots.yaml`, i.e. drop the `harness` part. I don't think it adds any value

egli replied:
Great, excellent

egli commented on tests/braille-specs/tr-g2_harness.yaml:
```diff
@@ -0,0 +1,312 @@
+  - ["I", ⠰⠠⠔]
+  - ["i", ⠰⠊]
+  - ["İ", ⠰⠠⠊]
+  - ["Punctuation simbols", ⠠⠏⠥⠝⠉⠞⠥⠁⠞⠊⠕⠝ ⠎⠊⠍⠃⠕⠇⠎]
```
Spelling: "simbols" should by "symbols"

hammera replied:
Thanks Chris, again my english. :-):-)
I fixed all testcase files from the simbols word with symbols word, sorry the typos.
Now I will be doing a force-push opreation, don't surprise, I uploading all requested modifications..
I think all requested thinks from your review comments related is resolved, except en-ueb-g1.ctb related include problem with tables/tr-g1.ctb table related, I am not have ydea how can possible resolve this request without bigger risks, near by the release date.
Perhaps better keeping now end of the table the include line, and the 3.30 release development cicle we have more time and more possibility to doing more optimizations the grade 1 table related, and fixing all 168 failures if en-ueb-g1.ctb line are removed with table level.
Because we have a proofread grade 1 test file with absolute known good, this is more easyer future (and more smaller the risk a three month development cicle). :-):-)
Into the 3.30 release development cicle Oguz already suggest more fixes (for example a match opcode related fix), so absolute sure need doing changes into the 3.30 development cicle with turkish grade 1 table.
I do not know how necessary the complex operations provided by the Turkish language table in the UEB (and how much the rules are required in the en-ueb-g1.ctb UEB table in the UEB Table subtable files are finally driven). So, it is better to test this change in a multi-month development cycle based on reading real Turkish-language documents when the base 168 defective test case is solved after the Include En-ueb-g1.ctb line has been removed.
If the last part is not full good with english language, sorry, I wrote this comment part from hungarian to english translation with Google translator. :-):-)

egli reviewed (changes requested):
Hi @hammera thanks for this great PR.

I have just two comments

1.  I prefer that the names of the yaml files to be just `tr-8dots.yaml`, i.e. drop the `harness` part. I don't think it adds any value
2. Why did you remove the `include` of the `en-ueb-g1.ctb`. Is it because of the latin capital i? Is there a way to achieve your goal and still use the include?

Excellent work, thank you very much

hammera commented:
Ok, I think now all changes looks good. :-):-)
All checks are passed, Chris, if this is possible please resolve already solved review requested changes comments conversions.
Now I think all changes is full final review ready.

Attila

hammera commented:
@egli, thanks the approve.
@bertfrees, if you have a little time have possibility to remove the needs test label this PR related?
All turkish tables related I doed test case files with @oguzugur28 are proof readed.
@egli, thanks the approve.
@bertfrees, if you have a little time have possibility to remove the needs test label this PR related?
All turkish tables related I doed test case files with @oguzugur28 are proof readed.

I wish everybody a good day,

Attila

bertfrees commented on tables/tr-g2.ctb:
```diff
@@ -156,74 +240,200 @@ digit 9 247
 # multind 56-6 letsign capsign
 
-# Latin Harfleri
+# Accented Letters
```
What concerns me a bit is that the grade 1 table is based on UEB, but the grade 2 isn't and takes definitions from the German table.

hammera replied:
@bertfrees wrote following he's review comment part:
> What concerns me a bit is that the grade 1 table is based on UEB, but the
> grade 2 isn't and takes definitions from the German table.

With original in 2017 Arend amend why designed turkish grade 1 braille system with ueb system, I don't known unfortunately. Technicaly and interests level I tried to look up the peculiarities of the Braille system in Turkish last week.
If you have guessed correctly, the Turkish Braille system is based on a mixed system, which includes Latin letters, accented characters based on the German language, but also some French and Irish characters. So a real mixed salad, at least according to the Wikipedia entry last modified in 2023:
[Turkish Braille](https://en.wikipedia.org/wiki/Turkish_Braille)
I not founds other good informations this topic related with helps understanding turkish Braille more advanced level or technical level to we detailed understanding the turkish level braille systems differences.

bertfrees commented on tables/tr.ctb:
```diff
@@ -20,26 +22,46 @@
+# 1991. For an unaccessible PDF version see
+# <https://orgm.meb.gov.tr/alt_sayfalar/yayimlar/ozelegitim/braille/braille.pdf>.
+#
+# Definitions of accented letters do not follow the guideline, but
+# were taken from de-accents.cti instead. This was done to prevent
+# mixing accented letters with contractions. For example, some
+# accented e letters should normally be represented by dot 26 or 1246,
+# but these patterns are also used for contractions.
```
There are no contractions in computer braille, so what could be mixed up?

hammera replied:
This is a good question, @oguzugur28, please answer Bert question because I am not known turkish language.

hammera replied:
Not possible Oguz simple wanted wrote braille word part?
Usual computer braille real not have contraction, @bertfrees, you wrote right, but unfortunately @oguzugur28 doesn't answered this question, and me not have other informations this question related.

bertfrees replied:
I've added a note in the table about this question, which I would like to be cleared up at some point, but it shouldn't block the PR.

bertfrees commented on tables/tr.ctb:
```diff
@@ -82,32 +104,222 @@ lowercase m 134		LATIN SMALL LETTER M
+base uppercase        \x1E84 \x1E85      Ẅẅ  &#7812;&#7813; LATIN LETTER W WITH DIAERESIS
+base uppercase        \x1EF2 \x1EF3      Ỳỳ  &#7922;&#7923; LATIN LETTER Y WITH GRAVE
+
+# Correct order of comma and numeric indicator
+match %a , %# 2-3456
+noback pass2 @3456-2-3456 @2-3456
+
+# Correct order of dot and numeric indicator
+match %a . %# 256-345
+noback pass2 @3456-256-3456 @256-3456
```
There is no numsign in this table, so why are these rules needed?

bertfrees replied:
I just deleted the rules, I don't think we need them. We'll notice if we do anyway.

bertfrees commented on tests/braille-specs/tr.yaml:
```diff
@@ -0,0 +1,870 @@
+  language: tr
+  type: computer
+  __assert-match: tr.tbl
+flags: {testmode: forward}
```
tr.tbl contains `#+direction:both` (it was marked this way because it is a computer braille table), however there areno tests for backward translation. If I change this line to `{testmode: bothDirections}`, a lot a tests fail.

bertfrees replied:
My question was mainly about the computer braille table, because usually computer braille tables provide a one-to-one mapping.

I think I would indeed prefer the tests that work in both directions, to run in both direction (only for the computer braille table). The ones that can't be back-translated because of conflicts could get a `{xfail: {backward: conflict}}` flag.

hammera replied:
Thanks Bert, ok, I think I understanding what you would like.
I will looking the computer braille entries the test case file and marks the conflicting parts the test case file with you wrote flag.
My question:
the three equals character into the test file are need? or an unwanted change before or after the table header part (I am not remember the proper place, but very similar a perhaps merge conflicting header part, when a conflict part marks three equals character an already putted part a file).
Another change I need doing before merge operation?

Attila

bertfrees replied:
I don't understand the question.

hammera replied:
I already doed the bothDirections change with computer Braille related testcase part.
Now I doing marking the braille dots conflicting back-translation failures with the {xfail: {backward: conflict}} flag.
Previous, when I created the testcase files I forgot lou_checkyaml already supports these failure markups, sorry, next time I using this technique.
Lot of conflicts are impossible resolve with table level, because the \x00e0, \x00e1 or \x00e2 characters are equals need marking I think with dots 4-1 braille dots.

Bert, I doing similar bothDirections mode change with grade1 and grade 2 parts the testcase, or enough the computer braille part into the test file?
I would like doing only one commit and one push the test file related. :-):-)

Attila

hammera replied:
Bert, I am suspicious turkish grade1 and grade2 tables is not back-translation compatible yet.
When I change forward with bothDirections mode, I get 200 failures with back-translation direction only the grade 1 table related, not all failures is dot conflicts, but very lot of dots are conflicting another braille equals dots, similar with computer Braille table.
Looks for example the uppercase A letter back-translation result, very interesting the result, because I am not known where from apply Liblouis a pass4 rule or an another rule with produce wrong output, only sure to I am not putted any pass2 rules the turkish grade 1 table. With a code part I putting lou_trace -b command result with unicode.dis,tr-g1.ctb command result:
```
tools/lou_trace unicode.dis,tr-g1.ctb -b
⠠⠁
◌a
1.	pass2	$sldpUumSDwxyz	*#1=0
2.	pass2	$d	*
3.	always	◌	6
4.	lowercase	a	1
```
What? Why? :-):-)
Turkish grade 1 table doesn't contains pass2 rules, perhaps this pass2 rules coming from en-ueb-g1.ctb or an another subtable with en-ueb-g1.ctb table are included.
The \x25cc character is need marking with dot 6,
I am not known visual the \x25cc character is a turkish letter?
Perhaps if we putting a noback prefix before with \x25cc character rule, possible resolve the situation with uppercase a related, but this is only one failure.  :-):-)
So:
The computer Braille related I marked back-translation conflicting test cases with you requested flag the testcase entries, and in computer braille section I changed the forward direction with bothDirections flag.
Grade1 test section I keeped the forward direction, because if I changing the direction with bothDirections mode, I get 200 failure results.
I not looked the grade2 results, but I am suspicious the failure result quality perhaps similar.
I suggesting because have only I think four day before release publication, now keeps forward direction with grade1 and grade2 testcases, and trying solve grade1 testcase bothDirections change the 3.30 development cicle, because the failure quality require more debugging work with back-translation direction.
You accept this compromise suggestion?
If yes, in the 3.30 cicle I trying debugging turkish grade 1 table and trying decrease the back-translation failure quality. :-):-)
If not accepts this compromise suggestion, I trying too debugging the 3.30 cicle the turkish grade1 table, because disturbing me the 200 failures with back-translation direction. :-):-)
This situation need marking in the test file the grade 1 section what test are conflicting xfail testcase, and what test are simple xfailing test with need fixing the future. :-):-)

Attila

bertfrees replied:
OK that is fine.

hammera replied:
This is interesting...
I examined the grade 2 bothDirection mode, the tests produces only 163 failures to back-translation direction, similar with computer Braille tests.
My original grade2 test file contains 286 test entries, this is not a bad percentage. :-):-)
So, in the 3.30 development cicle absolute sure need improve tr-g1.ctb table back-translation quality, because my original grade 1 file contains 255 test entries, and 200 entries are back-translation failing. This is horrible quality for back-translation direction, forward direction is fine.
Perhaps the 3.30 cicle entire need redesigning tr-g1.ctb table, and need remove the last en-ueb-g1.ctb include line.
But I don't understand a think, perhaps future need consulting Arend Amend, he's the original table author I think in 2017 if I remember correct from commit history:
Original table creation time Arend why designed the tr-g1.ctb table to include en-ueb-g1.ctb table? Why very need this include line?
Only latin letters not require this include line, sure have purpose this, not an unvanted include operation happened by design (Arend Amend I think an experienced table author).
I see already complicated complex Braille systems, similar with hungarian language (hungarian language braille systems is relative complex, turkish braille is absolute more complex), but the turkish braille system is an interesting „mix” for braille related (contains english braille systems rules, but contains german and french braille systems based rules, and few irish language based rules). :-):-)
I understanded this from this very fresh Wikipedia article when I worked table fixes and want little learning with turkish braille to I understanding what and why need doing a required way:
[Turkish Braille Wikipedia Article](https://en.wikipedia.org/wiki/Turkish_Braille)

So:
Now I marking conflicting back-translation entries with the proper flag the grade 2 test case entries, similar with computer braille tests already we doed, and change grade2 test direction with bothDirections flags.
Perhaps now we keeping the grade 1 tests with forward direction, and after 3.29.0 release publication we redesigning the grade1 table, together with Arend Amend if he subscribed the Liblouis list.
Forward direction presenting is fine, need improve the back-translation quality if this is possible, except the equals dot combinations related rules (impossible to fixing this type rules with back-translation direction, because this rules uses equals dot combinations).

Attila

hammera replied:
Oh... I am possible was little tired when I added the \x25cc symbol the tables/tr-g1.ctb table, simple I forgotted putting the noback prefix before always opcode (now detected this). :-):-)
After I doed this change, the 200 failures in grade 1 both Directions mode decreased with 136 failures, and the 136 failures are usual with braille dots conflicting failures, similar with other turkish 8 dots and grade2 tables).
Now I marks conflicting failures the grade 1 tests the proper flags, and do a final push when I am ready.
When I worked the grade 1 table update, @oguzugur28 sent me a large character and correct list (I think the list have more than 141 rules), when comed the evening, I seed all place with unicode hexa codes with all places. :-):-)
So, the final conclusion:
* All tables, all test case parts are bothDirections mode,
* All test sections I will be marking with conflicting xfail entries. Unfortunatelly with future this test case entries impossible to resolve back-translation direction the failures, because all characters uses equals dot combinations, bot forward and backward direction. With turkish language if I known right not have special braille input prefix, for example with dot 5 when a dot combination need doing different meaning with back-translation direction (a typical example the hungarian 1456-1456, and 1456-5-1456 with back-translation direction. If back-translation the hungarian user type two 146 dot combination he's perkins keyboard, need back-translate ccs text, if type for example 1456-5-1456, need back-translate cscs text). With turkish language not awailable this type rules and the standard are not defining this type rules.

Boys, I think we doed everiting with turkish computer braille, grade 1 braille and grade 2 braille with this is possible. The final result will be not too bad. So, in finish, have the happy end! :-):-)
So, I am finishing the grade 1 test failures markup, and do a final push.
After this, the new committed changes are review ready I think.
Sorry the previous not detected forgotted noback prefix, this is already fixed with tables/tr-g1.ctb, but not pushed yet.

Attila

hammera replied:
Ok, I changed grade 1 tests with bothDirections mode, and marked the conflicting entries with following flag:
{xfail: {backward: conflict}}
I remember wrong number with grade 1 tests, from 255 or 256 tests after the noback related previous change only 133 tests are failed.
So, this development cicle I done the turkish table related work, entire yesterday and this day added changes are review ready my side.
Both three Bert doed test case sections the tests/braille-specs/tr.yaml file are bothDirections mode, all test sections are marked the conflicting back-translation xfail entryes.

I wish everybody a good weekend and good rest if this is possible, tomorrow I will be limited near by internet, sunday entire day I am not in home.

Everybody I wishing a good 3.29.0 Liblouis release in next week monday. :-):-)

Thanks the cooperation during review,

Attila

hammera replied:
The third conversion part with test files related are possible resolve, all tasks done with test files related.
I will quote other comment part from @bertfrees the proper sections.
Usual this parts are technical questions, with me unable to answering, because native I am not known turkish language.
So, if everibody accepts this, I am or perhaps Chris or Bert resolve the test files related conversion part to easying see remaining questions and tasks.

bertfrees reviewed (changes requested):
See my comment.

hammera commented:
@bertfrees,  without turkish language knowledge tomorow morning what can I have possibility to fixing from you requested changes list?
So, it would be good for me to make changes that do not require me to know the Turkish language. Oguz will be able to answer you regarding the match rules and differences related to accents (mixed latin and german accents with you founded the review). :-):-)

hammera commented:
Boys, sorry the again rebase and force-push operation, only I edited with two commit messages, so I fixed few horrible typing errors my longer documentation commit messages. :-):-)
What other remaining review requested changes are need addressing before merge?
Bert asked some questions with he's review with I unable to answering unfortunately (I think with tr.ctb or tr-g1.ctb, perhaps with tr-g2.ctb tables related with different accents characters related).
Have opened with 4 conversions with need resolve before merge operation.
The test files related entire changes task we resolved I think last week (changing computer tests with bothDirections mode and marks conflicting test with a proper xfail flag), change grade 1 tests with bothDirections mode and marks conflicting back-translation tests with conflicting xfail flag, change grade2 tests with bothDirections mode and marks with , conflicting back-translation xfail flag), so this task part are adressed and fixed.
I forgot last week doing an another requested change? :-):-)
I will looking Bert longer review comments again, and will be quote perhaps what changes are remaining and possible assigning and fixing this last day.
All check fortunately rans succesfull, so technical everithing looks good to merge related.
All not conflicting or simple xfail marked turkish tests are executed correct both forward and backward directions, all turkish tables (tables/tr.ctb, tables/tr-g1.ctb, and tables/tr-g2.ctb), so this is a very good result, if we take into account the fact that we have only been working on the entire pull request since February 14th.
Although it should be mentioned that, as far as I know, with Oguz and he's friends have invested a lot of work in improving the tr.ctb, tr-g1.ctb and tr-g2.ctb boards from 2023. So, I think there is reason to be happy, this is a nice result, beautiful work. :-):-)
After Oguz learning best techniques more level, will be a fantastic maintainer member with turkish language related my openion.

Attila

bertfrees commented on tests/braille-specs/tr.yaml:
```diff
@@ -0,0 +1,870 @@
+  - ["İ", ⠠⠊]
+
+  # sentences
+  - ["(Örnek bir metin)", ⠶⠠⠪⠗⠝⠑⠅ ⠃⠊⠗ ⠍⠑⠞⠊⠝⠶, {xfail: {backward: true}}]
```
This test fails after I rebased the branch onto master. In commit 192a04496b375203f51b4bf6655f5f1898776c1e, @jrbowden added the following rule:

    postpunc ) 5-345

The explanation is

> add prepunc and postpunc definitions for several signs to prevent regression backtranslating words followed by punctuation.

So I'm not sure if this was supposed to influence forward translation.

Anyway, I'm not sure what to do with this test.

hammera replied:
I trying fixing this, readed your review now.
I seed Other @jrbowden doed changes with en-ueb-g1.ctb table.
Anyway, in turkish tables absolute sure need marks 2356 the ( and ) character in grade 1, so perhaps next development cycle better remove en-ueb-g1.ctb table if technical not have a larger need why need include this file by design. :-):-)
Me need syncing the update_turkish_tables branch my local system.

Attila

bertfrees replied:
For now I marked it with `xfail`.

bertfrees replied:
@hammera I've force-pushed all my changes just now. Please synchronize again.

hammera replied:
Ok, thanks the notification. I synced again. :-):-)
Understanding what the situation with tr-g1.ctb table level with forward direction, looks following lou_trace example:
```
(alma)
⠶⠁⠇⠍⠁⠐⠜
1.	punctuation	(	2356
2.	lowercase	a	1
3.	lowercase	l	123
4.	lowercase	m	134
5.	lowercase	a	1
6.	postpunc	)	5-345
```
This is easy fixable with following patch (I tested this) now, but not committed yet, because need modifying the ( and ) related test example the grade 1 section in to the 549 line:
```
diff --git a/tables/tr-g1.ctb b/tables/tr-g1.ctb
index b0848ed7..c1ccbc9d 100644
--- a/tables/tr-g1.ctb
+++ b/tables/tr-g1.ctb
@@ -237,7 +237,9 @@ noback always \x25cc 6
 punctuation ' 3
 punctuation ; 23
 punctuation : 25
+prepunc ( 2356
 punctuation ( 2356
+postpunc ) 2356
 punctuation ) 2356
 punctuation − 36
 
diff --git a/tests/braille-specs/tr.yaml b/tests/braille-specs/tr.yaml
index b0548b0a..502b4a44 100644
--- a/tests/braille-specs/tr.yaml
+++ b/tests/braille-specs/tr.yaml
@@ -546,9 +546,7 @@ tests:
   - ["İ", ⠠⠊]
 
   # sentences
-  - ["(Örnek bir metin)", ⠶⠠⠪⠗⠝⠑⠅ ⠃⠊⠗ ⠍⠑⠞⠊⠝⠶,
-     {xfail: {forward: fails since commit 192a04496b (with addition of "postpunc ) 5-345" rule),
-              backward: true}}]
+  - ["(Örnek bir metin)", ⠶⠠⠪⠗⠝⠑⠅ ⠃⠊⠗ ⠍⠑⠞⠊⠝⠶]
   - ["Bu nasıl olabilir?", ⠠⠃⠥ ⠝⠁⠎⠔⠇ ⠕⠇⠁⠃⠊⠇⠊⠗⠦]
   - ["Bu nasıl olabilir…", ⠠⠃⠥ ⠝⠁⠎⠔⠇ ⠕⠇⠁⠃⠊⠇⠊⠗⠲⠲⠲]
   - ["\"En büyük padişah Sultan I. Süleyman'dır.\"",```
```
This fixing way back-translation part work too. :-):-)

And this is the right wanted output the grade 1 forward direction in turkish braille:
```
(alma)
⠶⠁⠇⠍⠁⠶
1.	prepunc	(	2356
2.	lowercase	a	1
3.	lowercase	l	123
4.	lowercase	m	134
5.	lowercase	a	1
6.	postpunc	)	2356
```
So, to winning the precedence rule order the need turkish prepunc and postpunc dot combination, need defining proper rules before en-ueb-g1.ctb include line.
Not beautiful the fix, but works and resolves the unwanted situation with the new en-ueb-g1.ctb table related change doed the turkish grade 1 system. :-):-)
I committing the change.

Attila

hammera replied:
Done, pushed the change, now running the required checks.
Me the make -j8 check command ran successfull the test directory my local system before pushing.
I drinking a coffee, and perhaps the remaining two online check running right until drinking the cofee too. :-):-)
So, after this change verify result, we ready this PR this development cicle, all situations are resolved.

Attila

hammera replied:
Nice, all checks passed. :-):-)

hammera commented:
@oguzugur28, congratulate the new tables.
Bert, Chris, I would like thank you the contribution and cooperation.

Attila

egli commented:
> @oguzugur28, congratulate the new tables. Bert, Chris, I would like thank you the contribution and cooperation.

Thank you Attila for all your hard work
